sockeye/inference.py

```diff
@@ -480,46 +554,48 @@ def _make_result(self,
 
     def translate_nd(self,
                      source: mx.nd.NDArray,
-                     bucket_key: int) -> Tuple[np.ndarray, np.ndarray, float]:
+                     source_length: int) -> Tuple[np.ndarray, np.ndarray, float]:
         """
         Translates source of source_length, given a bucket_key.
 
         :param source: Source ids. Shape: (1, bucket_key).
-        :param bucket_key: Bucket key.
+        :param source_length: Bucket key.
 
         :return: Sequence of translated ids, attention matrix, length-normalized negative log probability.
         """
-        # allow output sentence to be at most 2 times the current bucket_key
-        # TODO: max_output_length adaptive to source_length
-        max_output_length = bucket_key * C.TARGET_MAX_LENGTH_FACTOR
+        return self._get_best_from_beam(*self._beam_search(source, source_length))
 
-        return self._get_best_from_beam(*self._beam_search(source, bucket_key, max_output_length))
-
-    def _encode(self, source: mx.nd.NDArray, bucket_key: int) -> List[ModelState]:
+    def _encode(self, source: mx.nd.NDArray, source_length: int) -> List[ModelState]:
         """
         Returns a ModelState for each model representing the state of the model after encoding the source.
 
         :param source: Source ids. Shape: (1, bucket_key).
-        :param bucket_key: Bucket key.
+        :param source_length: Bucket key.
         :return: List of ModelStates.
         """
-        prev_target_word_id = mx.nd.full((self.beam_size,), val=self.start_id, ctx=self.context)
-        model_states = [ModelState(bucket_key=m.encoder.get_encoded_seq_len(bucket_key),
-                                   prev_target_word_id=prev_target_word_id,
-                                   decoder_states=m.run_encoder(source, bucket_key))
-                        for m in self.models]
-        return model_states
+        return [ModelState(states=m.run_encoder(source, source_length)) for m in self.models]
 
-    def _decode_step(self, states: List[ModelState]) -> Tuple[mx.nd.NDArray, mx.nd.NDArray, List[ModelState]]:
+    def _decode_step(self,
+                     sequences: mx.nd.NDArray,
+                     t: int,
+                     bucket_key: Tuple[int, int],
+                     states: List[ModelState]) -> Tuple[mx.nd.NDArray, mx.nd.NDArray, List[ModelState]]:
         """
         Returns decoder predictions (combined from all models), attention scores, and updated states.
 
         :param: List of model states.
         :return: (probs, attention scores, list of model states)
         """
+        # bucket target max length based on beam_search progress
+        if len(self.buckets_target) > 1:
+            target_max_length = data_io.get_bucket(t, self.buckets_target)
+            if target_max_length < bucket_key[1]:
+                bucket_key = bucket_key[0], target_max_length
```

Comment:
Thats probably a question of taste. I prefer not having this logic in the beam search to not make this overly complicated and let the decode step decide how to effectively pass data to the module.

Reply:
yeah sure. But then I'd rather just pass in the source length and create the bucket_key inside the method, rather than conditionally overwriting the second element of the key.

Reply:
But then you also need to pass in max_output_length at which point you almost pass in the default bucket key again :)
If you feel strongly about this I can pull the logic out of decode_step but otherwise I'ld leave as is.

Reply:
I don't feel strongly about this, but I do think it would be cleaner to just pass in max_output_length as well, because after all what's called `bucket_key` is really only a bucket_key when no bucketing is used, otherwise it only carries information about the source length.